xclim/indices/_agro.py

```diff
@@ -882,6 +883,188 @@ def water_budget(
+    freq : str
+      Resampling frequency.
+
+    Returns
+    -------
+    rain_season_start: xr.DataArray, [dimensionless]
+    rain_season_end: xr.DataArray, [dimensionless]
+    rain_season_length: xr.DataArray, [time]
+
+    Notes
+    -----
+    The rain season starts at the end of a period of raining (a total precipitation  of `thresh_wet_start` over `window_wet_start` days). This must
+    be directly followed by a period of `window_not_dry_start` days with no dry sequence. The dry sequence is a period of `window_dry_start`
+    days where precipitations are below `thresh_dry_start` (either the total precipitations over the period, or the daily precipitations, depending
+    on `method_dry_start`). The rain season stops when a dry sequence happens (the dry sequence is defined as in the start sequence,
+    but with parameters `window_dry_end`, `thresh_dry_end` and `method_dry_end`). The dates on which the season can start are constrained
+    by `date_min_start`and `date_max_start` (and similarly for the end of the season).
+
+    References
+    ----------
+    :cite:cts:`sivakumar_predicting_1998`
+    """
+    # Unit conversion.
+    pram = rate2amount(pr, out_units="mm")
+    thresh_wet_start = convert_units_to(thresh_wet_start, pram)
+    thresh_dry_start = convert_units_to(thresh_dry_start, pram)
+    thresh_dry_end = convert_units_to(thresh_dry_end, pram)
+
+    last_doy = pram.indexes["time"][-1].strftime("%m-%d")
+
+    # Eliminate negative values.
+    pram = xarray.where(pram < 0, 0, pram)
+    pram.attrs["units"] = "mm"
+
+    # should we flag date_min_end  < date_max_start?
+    def _get_first_run(run_positions, start_date, end_date):
+        run_positions = select_time(run_positions, date_bounds=(start_date, end_date))
+        first_start = run_positions.argmax("time")
+        return xarray.where(
+            first_start != run_positions.argmin("time"), first_start, np.NaN
+        )
+
+    # Find the start of the rain season
+    def _get_first_run_start(pram, window_dry=window_dry_start):
+        pram = select_time(pram, date_bounds=(date_min_start, last_doy))
+
+        # First condition: Start with enough precipitation
+        da_start = pram.rolling({"time": window_wet_start}).sum() >= thresh_wet_start
+
+        # Second condition: No dry period after
+        if method_dry_start == "per_day":
+            da_stop = pram <= thresh_dry_start
+        elif method_dry_start == "total":
+            da_stop = pram.rolling({"time": window_dry}).sum() <= thresh_dry_start
+            # equivalent to rolling forward in time instead, i.e. end date will be at beginning of dry run
+            da_stop = da_stop.shift({"time": -(window_dry - 1)}, fill_value=False)
+            window_dry = 1
+
+        # First and second condition combined in a run length
+        run_positions = (
+            rl.rle_events(da_start, 1, da_stop, window_dry, "first")
+            >= window_not_dry_start + window_wet_start
```

Comment:
I found it weird that `window_dry` was given in the signature.
```suggestion
    def _get_first_run_start(pram):
        pram = select_time(pram, date_bounds=(date_min_start, last_doy))

        # First condition: Start with enough precipitation
        da_start = pram.rolling({"time": window_wet_start}).sum() >= thresh_wet_start

        # Second condition: No dry period after
        if method_dry_start == "per_day":
            da_stop = pram <= thresh_dry_start
            window_dry = window_dry_start
        elif method_dry_start == "total":
            da_stop = pram.rolling({"time": window_dry_start}).sum() <= thresh_dry_start
            # equivalent to rolling forward in time instead, i.e. end date will be at beginning of dry run
            da_stop = da_stop.shift({"time": -(window_dry_start - 1)}, fill_value=False)
            window_dry = 1

        # First and second condition combined in a run length
        run_positions = (
            rl.rle_events(da_start, 1, da_stop, window_dry, "first")
            >= window_not_dry_start + window_wet_start
```